cmd/coredns-dnscache/main.go

```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
+	"fmt"
+
+	"flag"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/coredns/coredns/coremain"
+	_ "github.com/coredns/coredns/plugin/bind"
+	_ "github.com/coredns/coredns/plugin/cache"
+	_ "github.com/coredns/coredns/plugin/errors"
+	_ "github.com/coredns/coredns/plugin/forward"
+	_ "github.com/coredns/coredns/plugin/health"
+	_ "github.com/coredns/coredns/plugin/log"
+	_ "github.com/coredns/coredns/plugin/loop"
+	_ "github.com/coredns/coredns/plugin/metrics"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	_ "github.com/coredns/coredns/plugin/reload"
+	"github.com/mholt/caddy"
+	"k8s.io/kubernetes/pkg/util/dbus"
+	"k8s.io/kubernetes/pkg/util/iptables"
+	utilexec "k8s.io/kubernetes/pkg/util/exec"
+)
+
+type configParams struct {
+	localIP        string
+	localPort      string
+	intfName       string
+	interval       time.Duration
+	exitChan       chan bool
+}
+
+type ruleStruct struct {
+	table iptables.Table
+	chain iptables.Chain
+	args  []string
+}
+
+var iptInterface iptables.Interface
+var params = configParams{exitChan: make(chan bool, 1), localPort : "53"}
+
+var iptablesRules []ruleStruct
+
+func init() {
+	err := parseAndValidateFlags()
+	if err != nil {
+		clog.Errorf("Error parsing flags - %s", err)
+		return
+	}
+	initIptables()
+
+	teardown(nil)
+	err = setup(true)
+	if err != nil {
+		clog.Errorf("Failed to setup - %s, Exiting", err)
+		return
+	}
+	caddy.OnProcessExit = append(caddy.OnProcessExit, func() { teardown(params.exitChan) })
+}
+
+func initIptables() {
+
+	iptablesRules = []ruleStruct{
+		// for packets destined to nodelocalcache
+		{iptables.Table("raw"), iptables.ChainPrerouting, []string{"-p", "tcp", "-d", params.localIP,
+			"--dport", params.localPort, "-j", "NOTRACK"}},
+		{iptables.Table("raw"), iptables.ChainPrerouting, []string{"-p", "udp", "-d", params.localIP,
+			"--dport", params.localPort, "-j", "NOTRACK"}},
+		{iptables.TableFilter, iptables.ChainInput, []string{"-p", "tcp", "-d", params.localIP, "--dport",
+			params.localPort, "-j", "ACCEPT"}},
+		{iptables.TableFilter, iptables.ChainInput, []string{"-p", "udp", "-d", params.localIP, "--dport",
+			params.localPort, "-j", "ACCEPT"}},
+		// for replies from nodelocalcache
+		{iptables.Table("raw"), iptables.ChainOutput, []string{"-p", "tcp", "-s", params.localIP,
+			"--sport", params.localPort, "-j", "NOTRACK"}},
+		{iptables.Table("raw"), iptables.ChainOutput, []string{"-p", "udp", "-s", params.localIP,
+			"--sport", params.localPort, "-j", "NOTRACK"}},
+		{iptables.TableFilter, iptables.ChainOutput, []string{"-p", "tcp", "-s", params.localIP, "--sport",
```

Comment:
It's probably worth a comment here saying that normally the output packets would be allowed by conntracking, but if we turn off conntracking then we have to explicitly allow them.  It bit me while I was playing around, so it'll likely help others to understand :-)

Reply:
Done